File: C++/bubble-sort.cpp
template<typename T>
void swap(T& element1, T& element2)
{
    T tmp = element1;
    element1 = element2;
    element2 = tmp;
}

template<typename T>
void bubbleSort(T*& arr, const int& size)
{
    bool sorted = false;

    for (int pass = 1; (pass < size) && !sorted; ++pass)
    {
        sorted = true;
        for (int index = 0; index < size - pass; ++index)
        {
            int nextIndex = index + 1;
            if (arr[index] > arr[nextIndex])
            {
                swap(arr[index], arr[nextIndex]);
                sorted = false;
            }
        }
    }
}

File: Golang/bubble-sort.go
package main

func bubbleSort(arr []int) {

	isSorted := false
	for i := 0; (i < len(arr)-1) && !isSorted; i++ {
		isSorted = true
		for j := i + 1; j < len(arr); j++ {
			if arr[i] > arr[j] {
				arr[i], arr[j] = arr[j], arr[i]
				isSorted = false
			}
		}
	}
}

func bubbleSortRecursive(arr []int, n int) {
	if n == 1 {
		return
	}

	for i := 0; i < n-1; i++ {
		if arr[i] > arr[i+1] {
			arr[i], arr[i+1] = arr[i+1], arr[i]
		}
	}

	bubbleSortRec(arr, n-1)
}